api/v1_coins_post_redeem.go

```diff
@@ -0,0 +1,234 @@
+package api
+
+import (
+	"crypto/ecdsa"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"api.audius.co/config"
+	"api.audius.co/solana/spl/programs/reward_manager"
+	oap_rewards "github.com/OpenAudio/go-openaudio/pkg/rewards"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/gagliardetto/solana-go"
+	"github.com/gofiber/fiber/v2"
+	"github.com/jackc/pgx/v5"
+)
+
+func getAttestationSignatureForRewardClaimAuthority(rewardClaim oap_rewards.RewardClaim, claimAuthorityKey *ecdsa.PrivateKey) (string, error) {
+	claimData, err := rewardClaim.Compile()
+	if err != nil {
+		return "", fmt.Errorf("failed to get attestation bytes: %w", err)
+	}
+
+	hash := crypto.Keccak256(claimData)
+
+	signatureBytes, err := crypto.Sign(hash, claimAuthorityKey)
+	if err != nil {
+		return "", fmt.Errorf("failed to sign hash: %w", err)
+	}
+
+	return "0x" + hex.EncodeToString(signatureBytes), nil
+}
+
+// TODO: Implement/replace once core reward generation PR is merged
+func getClaimAuthority(mint solana.PublicKey) (*ecdsa.PrivateKey, error) {
+	return nil, nil
+}
+
+type CoinRewardParams struct {
+	Mint   string
+	Amount uint64
+}
+
+type RewardCodeRow struct {
+	Mint          string `db:"mint"`
+	RewardAddress string `db:"reward_address"`
+	Amount        uint64 `db:"amount"`
+	IsUsed        bool   `db:"is_used"`
+}
+
+func (app *ApiServer) v1CoinsPostRedeem(c *fiber.Ctx) error {
+	mintString := c.Params("mint")
+	if mintString == "" {
+		return fiber.NewError(fiber.StatusBadRequest, "Mint parameter is required")
+	}
+
+	// Read optional code from route params
+	redeemCode := c.Params("code")
+
+	myId := app.getMyId(c)
+	if myId == 0 {
+		return fiber.NewError(fiber.StatusBadRequest, "Missing user ID")
+	}
+
+	var userWalletAddress string
+	var userHandle string
+	err := app.pool.QueryRow(c.Context(), `SELECT wallet, handle FROM users WHERE user_id = @user_id`, pgx.NamedArgs{
+		"user_id": myId,
+	}).Scan(&userWalletAddress, &userHandle)
+	if err != nil {
+		if errors.Is(err, pgx.ErrNoRows) {
+			return fiber.NewError(fiber.StatusInternalServerError, "Failed to get user wallet")
+		}
+		return err
+	}
+
+	mint, err := solana.PublicKeyFromBase58(mintString)
+	if err != nil {
+		return fmt.Errorf("Mint is invalid: %s", mintString)
+	}
+
+	var rewardStateAddress string
+	var minVotes int
+	err = app.pool.QueryRow(c.Context(), `SELECT reward_manager_state, min_votes FROM sol_reward_manager_inits WHERE mint = @mint`, pgx.NamedArgs{
```

Comment:
Shortcut here to get min_votes from the table. Technically this could fail if the table has a different value than what's currently on chain. But that feels highly unlikely.